src/Protocol.cpp

```diff
@@ -0,0 +1,290 @@
+    point.y() = distance * vertical_angle_cos * sin(toRad(horizontal_angle_degree));
+    point.z() = distance * vertical_angle_sin;
+    return point;
+}
+
+base::Vector4d colorByReflectivity(uint8_t intensity)
+{
+    if (intensity < 30) {
+        return base::Vector4d(0,
+            static_cast<int>(intensity * 255 / 30) & 0xff,
+            0xff,
+            0xff);
+    }
+    else if (intensity < 90) {
+        return base::Vector4d(0,
+            0xff,
+            static_cast<int>((90 - intensity) * 255 / 60) & 0xff,
+            0xff);
+    }
+    else if (intensity < 150) {
+        return base::Vector4d(static_cast<int>((intensity - 90) * 255 / 60) & 0xff,
+            0xff,
+            0,
+            0xff);
+    }
+    else {
+        return base::Vector4d(0xff,
+            static_cast<int>((255 - intensity) * 255 / (256 - 150)) & 0xff,
+            0,
+            0xff);
+    }
+}
+
+std::optional<base::samples::Pointcloud> Protocol::handleSingleEcho(unsigned char* data)
+{
+    if (m_full_frame == true) {
+        m_point_cloud.points.clear();
+        m_point_cloud.colors.clear();
+        m_full_frame = false;
+    }
+    uint64_t timestamp_microseconds =
+        (static_cast<uint64_t>(data[1200] << 24) +
+            static_cast<uint64_t>(data[1201] << 16) +
+            static_cast<uint64_t>(data[1202] << 8) + data[1203]);
+    int miliseconds = timestamp_microseconds / 1000;
+    int microseconds = timestamp_microseconds % 1000;
+    base::Time time = base::Time::fromTimeValues(m_configuration.time_utc.year,
+        m_configuration.time_utc.month,
+        m_configuration.time_utc.day,
+        data[1197],
+        data[1198],
+        data[1199],
+        miliseconds,
+        microseconds);
+    for (int i = 0; i < 1197; i = i + 7) {
+        if (data[i] == 0xff && data[i + 1] == 0xaa && data[i + 2] == 0xbb &&
+            data[i + 3] == 0xcc && data[i + 4] == 0xdd) {
+            m_full_frame = true;
+            // TODO: Use the syncronized time here. We are using the time::now() at this
+            // moment cause we need a hardware integration between the lidar and a $GPRMC
+            // sender
+            m_point_cloud.time = base::Time::now();
+            return m_point_cloud;
+        }
+        if (data[i] < 128) {
+            double distance =
+                (data[i + 3] * 65536 + data[i + 4] * 256 + data[i + 5]) / 256.0 / 100;
+            if (distance != 0) {
+                uint8_t line_number = data[i];
+                double horizontal_angle_degree =
+                    (data[i + 1] * 256 + data[i + 2]) / 100.f;
+
+                uint8_t intensity = data[i + 6];
+
+                auto point = getPoint(line_number, horizontal_angle_degree, distance);
+                m_point_cloud.points.push_back(point);
+                m_point_cloud.colors.push_back(colorByReflectivity(intensity));
+            }
+        }
+    }
+    return std::nullopt;
```

Comment:
This is less verbose
```suggestion
    return {};
```